[ADD][8.0] hr_expense_product_policy

andhit-r opened the pull request:
Revision todo list:
- [x] change mapped("id") into .ids
- [x] assign recordset instead of list IDs on m2m computed field
- [x] domain on inherited view instead override fields_view_get
- [x] store or not field required_expense_product on hr.expense?
- [x] hard enforce policy on create and write method
- [x] Use other than Warning to raise exception
- [x] Remove unnecessary m2m field parameter
- [x] Remove unnecessary default=False on field creation
- [x] Split python file
- [x] @api.constraint on hr.expense
- [x] Remove empty lines in the middle
# 
# HR Expense Product Policy

This module add feature to set up policy regarding product on expense. Product
policy that ruled by this module are:
1. Force use of product on Expense
2. Allowed product that can be used based on employee's department
3. Allowed product that can be used based on employee's job
4. Allowed product that can be used based on employee himself
# Configuration

**Force Product Usage**
To force product usage on expense, you need to:
1. Open employee data
2. Open _HR Setting_ tab
3. Activate _Required Expense Product_

Hint:
You can use mass editing module to apply this rule

**Department's Expense Product Policy**
To set up department's allowed product policy based on product category, you need to:
1. Open department data
2. Select product category on _Allowed Expense Product Category_

To set up department's allowed product policy based on specific product, you need to:
1. Open department data
2. Select product on _Allowed Expense Product_

Note:
- Only product with _Can be expense_ flag turn on can be used, even if that product belong to category specified on _Allowed Expense Product Category_
- Only product with _Can be expense_ flag turn on can be selected on _Allowed Expense Product_
- Final result of allowed product can be seen on _All Allowed Expense Products_

**Job's Expense Product Policy**
To set up job's allowed product policy based on product category, you need to:
1. Open job data
2. Select product category on _Allowed Expense Product Category_

To set up job's allowed product policy based on specific product, you need to:
1. Open job data
2. Select product on _Allowed Expense Product_

Note:
- Only product with _Can be expense_ flag turn on can be used, even if that product belong to category specified on _Allowed Expense Product Category_
- Only product with _Can be expense_ flag turn on can be selected on _Allowed Expense Product_
- Final result of allowed product can be seen on _All Allowed Expense Products_

**Employee's Expense Product Policy**
To set up employee's allowed product policy based on product category, you need to:
1. Open employee data
2. Select product category on _Allowed Expense Product Category_

To set up employee's allowed product policy based on specific product, you need to:
1. Open employee data
2. Select product on _Allowed Expense Product_

Note:
- Only product with _Can be expense_ flag turn on can be used, even if that product belong to category specified on _Allowed Expense Product Category_
- Only product with _Can be expense_ flag turn on can be selected on _Allowed Expense Product_
- Final result of allowed product can be seen on _All Allowed Expense Products_. The list will be the aggregation of (1) department's product policy, (2) job's product policy, and (3) employee's product policy

mikevhe18 committed:
[ADD] Adding module hr_expense_required_product_by_employee

andhit-r committed:
(1) change name into hr_expense_product_policy, (2) finishing, (3) add unittest, and (4) add proper README

andhit-r commented:
Hi @OCA/human-resources-maintainers , would you kindly put need review label to this PR. Many thanks.

andhit-r commented:
Hi @OCA/human-resources-maintainers , would you kindly put need review label to this PR. Thanks in advance

dreispt commented on hr_expense_product_policy/models/hr_employee.py:
```diff
@@ -0,0 +1,166 @@
+        for employee in self:
+            products = employee.allowed_expense_product_ids
+            category_ids = employee.allowed_expense_product_categ_ids.mapped(
+                "id")
```
I believe that you can simply use`.ids`:

```
category_ids = employee.allowed_expense_product_categ_ids.ids
```

andhit-r replied:
Wow ... this is something i don't know. Thank you so much @dreispt. Will change it on my next push.

dreispt commented on hr_expense_product_policy/models/hr_employee.py:
```diff
@@ -0,0 +1,166 @@
+                products += job.all_allowed_expense_product_ids
+
+            employee.all_allowed_expense_product_ids = \
+                products.mapped("id")
```
You should assign the actual recordset rather that the list of IDs.

andhit-r replied:
You are right again. Would you kindly tell me the different? Because the test did not fail when i use list of IDs. Why using actual recordset better? Thanks in advance.

dreispt replied:
I don't why it worked before.
But I know that with the new API we should work with Record Objects instead of Record IDs.

dreispt commented on hr_expense_product_policy/models/hr_employee.py:
```diff
@@ -0,0 +1,166 @@
+        for dept in self:
+            products = dept.allowed_expense_product_ids
+            category_ids = dept.allowed_expense_product_categ_ids.mapped(
+                "id")
```
`.ids`?

dreispt commented on hr_expense_product_policy/models/hr_employee.py:
```diff
@@ -0,0 +1,166 @@
+        for job in self:
+            products = job.allowed_expense_product_ids
+            category_ids = job.allowed_expense_product_categ_ids.mapped(
+                "id")
```
`.ids`?

dreispt commented on hr_expense_product_policy/models/hr_expense.py:
```diff
@@ -0,0 +1,91 @@
+    required_expense_product = fields.Boolean(
+        string="Required Expense Product",
+        compute="_compute_required_expense_product",
+        store=True,
```
Do you really need to store this?

dreispt commented on hr_expense_product_policy/models/hr_expense.py:
```diff
@@ -0,0 +1,91 @@
+    @api.model
+    def fields_view_get(self, view_id=None, view_type='form', toolbar=False,
+                        submenu=False):
+        """Inject the domain here to avoid conflicts with other modules.
```
Can you explain a little better why you prefer not to add the domain directly to the view field?

andhit-r replied:
I want to avoid conflict if other module add/modify domain on product field. Do you have concern that i should consider when overriding fields_view_get?

I got the example from https://github.com/odoomrp/odoomrp-wip/tree/8.0/sale_order_allowed_product and https://github.com/odoomrp/odoomrp-wip/tree/8.0/purchase_order_allowed_product.

dreispt replied:
My concern is that it adds a lot of complexity for little gain.
You're probably anticipating problems that might not happen at all.
I suggest you to instead just extend the XML view.

dreispt commented on hr_expense_product_policy/models/hr_expense.py:
```diff
@@ -0,0 +1,91 @@
+                field = view["fields"][field_name]
+                domain = ustr(field.get("domain", "[]"))
+                field["domain"] = domain[:1] + domain_field + domain[1:]
+        return res
```
You could hard enforce the domain on the "write" method, so that if even something goes wrong with the client side domain, the user will get an error if the policy is not followed.

andhit-r replied:
Will do on my next push

dreispt commented on hr_expense_product_policy/models/hr_expense.py:
```diff
@@ -0,0 +1,91 @@
+    required_expense_product = fields.Boolean(
+        string="Required Expense Product",
+        compute="_compute_required_expense_product",
+        store=True,
```
It's simpler to make this a "related" field.
This is only needed if you add the domain to the form field, instead of overriding `fields_view_get`.

dreispt commented on hr_expense_product_policy/tests/__init__.py:
```diff
@@ -0,0 +1,5 @@
+# © 2016 OpenSynergy Indonesia
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from . import test_hr_expense
```
Your tests only [miss](https://travis-ci.org/OCA/hr/jobs/142426834#L1535) 6 lines of code.
Not bad!

dreispt commented on hr_expense_product_policy/views/hr_employee_views.xml:
```diff
@@ -0,0 +1,67 @@
+    <field name="inherit_id" ref="hr.view_employee_form"/>
+    <field name="arch" type="xml">
+        <data>
+            <xpath expr="//field[@name='active']" position="after">
```
Isn't this better inside a Notebook Page?

dreispt commented on hr_expense_product_policy/views/hr_expense_views.xml:
```diff
@@ -0,0 +1,29 @@
+                        <field name="required_expense_product" invisible="1"/>
+                    </xpath>
+                    <xpath expr="//field[@name='line_ids']/tree/field[@name='product_id']" position="attributes">
+                        <attribute name="attrs">{'required':[('required_expense_product','=',True)]}</attribute>
```
Already asked, but I wonder why we can't also add the domain here...

dreispt commented:
Functional review: look like an interesting feature.
Technical review:  good overall, only a few comments to tackle

andhit-r committed:
Modification according @dreispt comment

dreispt commented on hr_expense_product_policy/models/hr_employee.py:
```diff
@@ -22,12 +22,11 @@ def _compute_all_allowed_product_ids(self):
                 ("categ_id", "in", category_ids),
                 ("hr_expense_ok", "=", True),
-                ]
+            ]
```
AFAICT both closing forms are PEP8 compliant.
I personally prefer the indented one, or even better, to have the closing symbol at the end of the block's last line of code.
But this is just a comment, there are no explicit guidelines to enforce here.

andhit-r commented:
@dreispt i addressed some of your comment on my last push. Will finish others ASAP. I made a checklist on the top of PR description regarding all your point. Thank you for your review.

dreispt commented on hr_expense_product_policy/models/hr_expense.py:
```diff
@@ -4,6 +4,7 @@
 
 from openerp import api, fields, models
 from openerp.tools import ustr
+from openerp.exceptions import Warning as UserWarning
```
AFAIK `Warning`is deprecated. Use `ValidationError` or  `UserError` instead.
I believe that you can avoid an extra import and use `raise models.ValidationError`

andhit-r committed:
Add additional information regarding limit product selection on README.rst

andhit-r commented:
@dreispt i have address all your comments. Would you kindly review it again. Many thanks.

ehahouimohssine commented:
Nice functionality. looks good

pedrobaeza commented on hr_expense_product_policy/models/hr_employee.py:
```diff
@@ -0,0 +1,163 @@
+    allowed_expense_product_categ_ids = fields.Many2many(
+        string="Allowed Expense Product Categories",
+        comodel_name="product.category",
+        relation="rel_allowed_product_categ_2_employee",
```
Let Odoo to give automatically the name of the table and fields, so remove these fields.

pedrobaeza replied:
Make the same for the rest of many2many

pedrobaeza commented on hr_expense_product_policy/models/hr_employee.py:
```diff
@@ -0,0 +1,163 @@
+
+    required_expense_product = fields.Boolean(
+        string="Required Expense Product",
+        default=False,
```
This is not needed

pedrobaeza commented on hr_expense_product_policy/models/hr_employee.py:
```diff
@@ -0,0 +1,163 @@
+    )
+
+
+class HrDepartment(models.Model):
```
Split each model in a Python file

pedrobaeza commented on hr_expense_product_policy/models/hr_expense.py:
```diff
@@ -0,0 +1,113 @@
+                # check required product
+                if not product_id and product_required:
+                    strWarning = "Product has to be filled"
+                    raise models.ValidationError(strWarning)
```
This should be put as an `@api.constrains` on the field `product_id`.

pedrobaeza commented on hr_expense_product_policy/tests/test_hr_expense.py:
```diff
@@ -0,0 +1,261 @@
+            "limit_product_selection": True,
+            "line_ids": [
+                (0, 0, line1)]})
+
```
Don't let all of these empty lines in the middle. PEP8 encourages not to do it. Indentation is enough to make the code readable.

andhit-r committed:
Remove unnecessary required parameter

andhit-r committed:
Split model into different py file

andhit-r committed:
Fix coding style

andhit-r committed:
Remove unnecessary paramater on m2m field

andhit-r committed:
Use @api.constraints to force validation

andhit-r commented:
@pedrobaeza thank you for your reviews. I addressed all your reviews. Would you kindly review it again.

@ehahouimohssine thank you for your review

pedrobaeza commented:
:+1:

andhit-r commented:
Hi @OCA/human-resources-maintainers would you kindly review this PR. Many thanks.

dreispt commented:
:+1:

andhit-r commented:
Thanks @dreispt @pedrobaeza @ehahouimohssine
